cosmo_tester/test_suites/snapshots/__init__.py

```diff
@@ -62,58 +278,260 @@ def wait_for_execution(client, execution, logger):
+    with manager.ssh() as fabric_ssh:
+        for deployment in manager.client.deployments.list():
+            path = DEPLOYMENT_ENVIRONMENT_PATH.format(
+                tenant='default_tenant',
+                name=deployment['id'],
+            )
+            logger.info(
+                'Checking deployment env for {name} was recreated.'.format(
+                    name=deployment['id'],
+                )
+            )
+            fabric_ssh.run('test -d {path}'.format(path=path))
+            logger.info('Deployment environment was recreated.')
+    if tenant:
+        logger.info('Found correct deployments for {tenant}'.format(
+            tenant=tenant,
+        ))
+    else:
+        logger.info('Found correct deployments.')
+
+
+@contextmanager
+def set_client_tenant(manager, tenant):
+    if tenant:
+        original = manager.client._client.headers[CLOUDIFY_TENANT_HEADER]
+
+        manager.client._client.headers[CLOUDIFY_TENANT_HEADER] = tenant
+
+    try:
+        yield
+    except:
+        raise
+    finally:
+        if tenant:
+            manager.client._client.headers[CLOUDIFY_TENANT_HEADER] = original
+
+
+def upload_test_plugin(manager, logger, tenant=None):
+    if tenant:
+        logger.info('Uploading test plugin for {tenant}'.format(
+            tenant=tenant,
+        ))
+    else:
+        logger.info('Uploading test plugin')
+    with set_client_tenant(manager, tenant):
+        manager.client.plugins.upload(TEST_PLUGIN_URL)
+    return get_plugin_list(manager, tenant)
+
+
+def get_plugin_list(manager, tenant=None):
+    with set_client_tenant(manager, tenant):
+        return [
+            (
+                item['package_name'],
+                item['package_version'],
+                item['distribution'],
+            )
+            for item in manager.client.plugins.list()
+        ]
+
+
+def get_deployments_list(manager, tenant=None):
+    with set_client_tenant(manager, tenant):
+        return [
+            item['id'] for item in manager.client.deployments.list()
+        ]
+
+
+def get_secrets_list(manager, tenant=None):
+    with set_client_tenant(manager, tenant):
+        return [
+            item['key'] for item in manager.client.secrets.list()
+        ]
+
+
+def get_nodes(manager, tenant=None):
+    with set_client_tenant(manager, tenant):
+        return manager.client.nodes.list()
+
+
+def _cluster(request, cfy, ssh_key, module_tmpdir, attributes, logger,
+             hello_count, install_dev_tools=True):
+
+    manager_types = [request.param, 'master']
+    hello_vms = ['notamanager' for i in range(hello_count)]
+    managers = [
+        MANAGERS[mgr_type](upload_plugins=False)
+        for mgr_type in manager_types + hello_vms
+    ]
+
+    cluster = CloudifyCluster.create_image_based(
+            cfy,
+            ssh_key,
+            module_tmpdir,
+            attributes,
+            logger,
+            managers=managers,
+            )
+
+    if request.param == '4.0.1':
```

Comment:
Can this reside somewhere in `cluster.py`, where the manager is created? Maybe add a `run_after_creation` interface? I don't like it here.